src/engine/controls/bpmcontrol.cpp

```diff
@@ -750,118 +784,114 @@ double BpmControl::getBeatMatchPosition(
         }
         // We are between the previous and next beats so we can try a standard
         // lookup of the beat length.
-        getBeatContextNoLookup(
-                dThisPosition,
-                dThisPrevBeat,
-                dThisNextBeat,
-                &dThisBeatLength,
+        getBeatContextNoLookup(thisPosition,
+                thisPrevBeat,
+                thisNextBeat,
+                &dThisBeatLengthFrames,
                 nullptr);
     }
 
     TrackPointer otherTrack = pOtherEngineBuffer->getLoadedTrack();
-    mixxx::BeatsPointer otherBeats = otherTrack ? otherTrack->getBeats() : mixxx::BeatsPointer();
+    mixxx::BeatsPointer otherBeats =
+            otherTrack ? otherTrack->getBeats() : mixxx::BeatsPointer();
 
     // If either track does not have beats, then we can't adjust the phase.
     if (!otherBeats) {
-        return dThisPosition;
+        return thisPosition;
     }
 
-    double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
+    mixxx::FramePos otherPosition = samplePosToFramePos(pOtherEngineBuffer->getExactPlayPos());
 
-    double dOtherPrevBeat = -1;
-    double dOtherNextBeat = -1;
-    double dOtherBeatLength = -1;
+    mixxx::FramePos otherPrevBeat = mixxx::kInvalidFramePos;
+    mixxx::FramePos otherNextBeat = mixxx::kInvalidFramePos;
+    mixxx::FrameDiff_t dOtherBeatLength = -1;
     double dOtherBeatFraction = -1;
-    if (!BpmControl::getBeatContext(
-                otherBeats,
-                dOtherPosition,
-                &dOtherPrevBeat,
-                &dOtherNextBeat,
+    if (!BpmControl::getBeatContext(otherBeats,
+                otherPosition,
+                &otherPrevBeat,
+                &otherNextBeat,
                 &dOtherBeatLength,
                 &dOtherBeatFraction)) {
-        return dThisPosition;
+        return thisPosition;
     }
 
     if (dOtherBeatLength == -1 || dOtherBeatFraction == -1) {
         // the other Track has no usable beat info, do not seek.
-        return dThisPosition;
+        return thisPosition;
     }
 
-    double dThisSampleRate = m_pBeats->getSampleRate();
     double dThisRateRatio = m_pRateRatio->get();
 
     // Seek our next beat to the other next beat
     // This is the only thing we can do if the track has different BPM,
     // playing the next beat together.
-    double thisDivSec = (dThisNextBeat - dThisPosition) /
-            dThisSampleRate / dThisRateRatio;
+    double thisDivSec =
+            (thisNextBeat - thisPosition) / m_sampleRate / dThisRateRatio;
 
     if (dOtherBeatFraction < 1.0 / 8) {
         // the user has probably pressed play too late, sync the previous beat
         dOtherBeatFraction += 1.0;
     }
 
     dOtherBeatFraction += m_dUserOffset.getValue();
-    double otherDivSec = (1 - dOtherBeatFraction) *
-            dOtherBeatLength / otherBeats->getSampleRate() / pOtherEngineBuffer->getRateRatio();
+    double otherDivSec = (1 - dOtherBeatFraction) * dOtherBeatLength /
```

Comment:
We need here to make use of the values we have received with a single call to getBeatContext.
This gurantees that the values used for calculations are matching.

I think we have room to refactor this a bit to be seconds independent and use the ratio compared to the master buffer instead, but that is out of the scope of this PR. Here we need to restore the old behavior.